streaming/format/base/file.py

```diff
@@ -0,0 +1,302 @@
+
+        Returns:
+            int: Disk usage.
+        """
+        # Collect this file's current cache usage, and the locality of each of its phases.
+        phase_locs = []
+        du = 0
+        for phase in self.phases:
+            if phase:
+                phase_du = phase.inventory_local(listing)
+                if phase_du:
+                    du += phase_du
+                    phase_loc = Locality.LOCAL
+                else:
+                    phase_loc = Locality.REMOTE
+            else:
+                phase_loc = Locality.DNE
+            phase_locs.append(phase_loc)
+        phase_locs = np.array(phase_locs, np.int64)
+
+        # From phase localities, determine phase evictions according to the keep policy.
+        phase_dels = self.stream.safe_keep_phases.get_phase_deletions(phase_locs)
+
+        # Apply any evictions.
+        du += self.evict_phases(phase_dels)
+
+        # Finally, return our local dir usage after evictions.
+        return du
+
+    def _unzip(self) -> int:
+        """Decompress zip phase, resulting in raw phase, maybe deleting zip phase.
+
+        Returns:
+            int: Delta disk usage, in bytes.
+        """
+        if not self.zip_phase:
+            raise RuntimeError('Wanted to unzip a shard, but required metadata is missing.')
+
+        # Read the zip phase data.
+        zip_filename = self.zip_phase.get_local_filename()
+        zip_data = open(zip_filename, 'rb').read()
+
+        # Mandatory cheap compressed phase size check.
+        if len(zip_data) != self.zip_phase.size:
+            raise ValueError(f'Compressed data does not match the expected size: ' +
+                             f'{len(zip_data):,} bytes vs {self.zip_phase.size:,} bytes.')
+
+        # Decompress.
+        raw_data = decompress(self.zip_algo, zip_data)
+
+        # Mandatory cheap decompressed phase size check.
+        if len(raw_data) != self.raw_phase.size:
+            raise ValueError(f'Decompressed data does not match the expected size: ' +
+                             f'{len(raw_data):,} bytes vs {self.raw_phase.size:,} bytes.')
+
+        # Save the raw phase data.
+        raw_filename = self.raw_phase.get_local_filename()
+        tmp_filename = raw_filename + '.tmp'
+        with open(tmp_filename, 'wb') as out:
+            out.write(raw_data)
+        os.rename(tmp_filename, raw_filename)
+
+        # Collect the locality of each phase.
+        if self.can_phase:
+            can_loc = Locality.LOCAL if self.can_phase.is_local() else Locality.REMOTE
+        else:
+            can_loc = Locality.DNE
+        phase_locs = np.array([Locality.LOCAL, Locality.LOCAL, can_loc], np.int64)
+
+        # Given localities and policy, determine phase evictions.
+        phase_dels = self.stream.safe_keep_phases.get_phase_deletions(phase_locs)
+
+        # Delete phases we don't want.
+        return len(raw_data) + self.evict_phases(phase_dels)
```

Comment:
wouldn't we want to subtract out the size of the evicted phases from the length of the raw data to correctly calculate cache usage? If i'm deleting some phases then that's a reduction in cache usage, right?

if this is correct (I may be misunderstanding) then there should be some other subtractions in this file too

Reply:
oh just looked at `phase.py`, looks like `evict` returns a negative number

Reply:
Yeah, my convention was du = absolute disk usage, ddu = delta disk usage (pos or neg)

Reply:
https://github.com/mosaicml/streaming/pull/585/commits/44fd90ef7e16e63ba117db74eb215bccae07aa8b